[WIP]: Track recently selected languages

demalex opened the pull request:
### Description

Implementation of enhancement #299 suggested by @cmyr. Every time user makes a language selection, the language is stored to `UserDefaults`. 

Using `UserDefaults` directly makes classes tightly coupled with it and makes it hard to maintain and test. To get rid of problems like this, `UserDefaults` was covered with additional abstraction. The idea is to keep settings provider in one place (`AppDelegate` for now) and inject it as dependency into other consumers. There are other places where `UserDefaults` is used directly and from my perspective it requires to be refactored, but such refactoring is out of this enhancement scope.

`SettingsProvider` is responsible for communication with `KeyValueStorage`. It executes save and fetch operations, while `SettingsProvidingProcessor` decorates `SettingsProvider ` and brings additional logic for save languages operation. Appropriate test cases also have been added.

### Screenshot

<img width="422" alt="screen shot 2018-10-16 at 8 39 39 pm" src="https://user-images.githubusercontent.com/1524024/47040201-fbca6100-d185-11e8-9081-618e1b932cb9.png">

cmyr commented:
Cool, thanks for this. Will have to give it a closer read, but looks nice and tidy at first glance, and thanks for the tests!

(if anyone else @xi-editor/xi-mac wants to take a look here that would be welcome!)

jansol commented on XiEditor/Storage/SettingsProvider.swift:
```diff
@@ -0,0 +1,36 @@
+import Foundation
+
+protocol SettingsProviding {
+    func fetchingRecentlyUsedLanguages() -> [String]
```
This should probably be called `fetchRecentlyUsedLanguages` ("fetching" sounds very weird to me).

demalex replied:
@jansol yes, `fetchRecentlyUsedLanguages` definitely sounds better 👍🏻

jansol commented on XiEditor/Storage/SettingsProvider.swift:
```diff
@@ -0,0 +1,36 @@
+    func saveRecentlyUsed(languages: [String])
+}
+
+class SettingsProvider: SettingsProviding {
```
This architecture, it seems needlessly convoluted. Yes, it looks a bit like MVC but that falls apart because in practice you can only access the `SettingsProvider` through the `SettingsProviderProcessor`. I'm not sure what benefits this separation is supposed to bring.

Also having `SettingsProvider` and `SettingsProviding` which are almost the same thing is confusing. I'd rather call the protocol `SettingsProvider` or even `ISettingsProvider` (to clearly indicate that it's an interface/protocol) and this class something like `SettingsStore`, since all it does is provide access to the KeyValueStore. Or you could probably remove this class entirely since it's just a very thin wrapper over a dictionary.

demalex replied:
@jansol to be honest I did not get what you meant by `MVC` and access the `SettingsProvider` through the `SettingsProviderProcessor`. Could you please explain what did you mean?

The idea behind `SettingsProvider` is that it responsible for only for interacting with storage without any additional logic while `SettingsProvidingProcessor` decorates `SettingsProvider` adding level of logic.

This solutions are flexible from my point of view, because `SettingsProvider` implementation can be easily replaced with new one (for example new implementation could fetch settings from `SQLite` instead of `UserDefaults`) and injected into `SettingsProvidingProcessor`. So, final configuration has new provider with new storage and with same processor that contains unchanged languages processing logic. As the project evolves, `UserDefaults` may not be enough and switching to some type of data base could be done quite easy.

What do you think? Does it worth to have such solution?

Yes, naming definitely should be improved. I would not use `ISettingsProvide` since it is not a best practice for naming protocols in Swift. Also I am trying to use this guideline during API design [API Design Guidelines](https://swift.org/documentation/api-design-guidelines/).

demalex commented on XiEditor/Storage/SettingsProvider.swift:
```diff
@@ -0,0 +1,36 @@
+        case recentlyUsedLanguages
+    }
+
+    private let keyValueStorage: KeyValueStorage
```
Hi @jansol, thanks for the comment! I have introduced `KeyValueStorage ` protocol in order to have ability easily switch implementations without touching client code.

First benefit of such approach is that It allows to choose whatever key-value storage is needed (UserDefaults, Dictionary etc.). It is a rare case to be honest, but still.
Second benefit goes when it is needed to mock and stub such storages in tests. The main disadvantage working with `UserDefaults` is that `UserDefaults` persists the data between tests launch. From my experience it is very easy to forget to remove user defaults suite or  some value and eventually find out that not related test cases are falling because of not properly cleaned up `UserDefaults`. So, it is very easy to mess it up.

Having such protocol allows to create mock `DictionaryStorage` and use it as storage in tests.

What do you think? Does it make sense?

jansol replied:
It makes sense for mocking, yes. But I didn't see any case where this *actually* is anything other than `DictionaryStorage` (and no other implementations either), hence my suspicions.

demalex replied:
@jansol yes, I agree. So, do we keep the protocol? I think it worth keeping it, at least for mocking purposes. I am open to any other suggestions if any.

demalex commented on XiEditor/Storage/KeyValueStorage.swift:
```diff
@@ -0,0 +1,17 @@
+//
+//  Created by Alex Demishkevych on 10/14/18.
+//  Copyright © 2018 Raph Levien. All rights reserved.
+//
```
@jansol Updated

demalex commented:
@cmyr @jansol I have updated license headers and changed components naming a bit. Also I have added some info for conversations above. Please, have a look. Thanks!

cmyr commented:
I've given this a read through, and I do see @jansol's point: it feels like there's maybe _one_ more layer of abstraction than is necessary here? I like the idea of using a protocol for the main interface, but I did find the overall inheritance chain tricky to follow, in a way that didn't seem _totally_ necessary.

For instance: if `SettingsProcessor` was combined with `SettingsStore`, I don't think this would make the code any harder to maintain?

demalex commented:
@cmyr Yes, it makes sense. The initial idea was just to separate store interaction logic and business logic, so they both could be modified independently. But we can definitely combine this two classes and it wont make the codebase any harder to maintain. Lets keep it simple at least at this stage of project.

demalex commented:
@cmyr @jansol Guys, I have removed `SettingsProcessor` as separate entity. Now there are only `SettingsStore` that conforms to `SettingsProvider` protocol. As we discussed with @jansol `KeyValueStorage` protocol is good to have for testing and mocking, so this protocol remains the same. Also, I have attached a screenshot to initial post. Please, have a look! Thank you!

jansol commented on XiEditor/EditViewController.swift:
```diff
@@ -850,23 +840,65 @@ class EditViewController: NSViewController, EditViewDataSource, FindDelegate, Sc
+
+    private func buildRecentlyUsedLanguagesSection(from languages: [String]) -> [NSMenuItem] {
+        var items: [NSMenuItem] = []
+        let titleItem = NSMenuItem(title: NSLocalizedString("Recently Used", comment: "Recently Used"), action: nil, keyEquivalent: "")
```
Might want to clarify the comment with something like "[Language]"

demalex replied:
@jansol I did not get your comment. Could you please explain?

jansol replied:
The comment is meant to provide context to translators so they don't need to look at (and understand!) the code when writing translations. (They should only need to work with the strings file)

Actually on second thought, "Language list subtitle" would be an even better description.

jansol commented:
So languages are supposed to show as in use in both "recently used" and the general list at the same time? That seems like reasonable behavior, but I don't see where the recently used one is marked as in use in the code. Missed when committing?

demalex commented:
@jansol Hmm, no in use language is supposed to be selected only in general list section. I am sorry, but I think screenshot is misleading and I even do not know how I made it. 
Do you think it makes sense to show language as selected in both sections?

jansol commented:
Yeah, I think it's more consistent that way.

demalex commented:
@cmyr @jansol Updated screenshot and implementation. Please, have a look!

demalex commented:
@jansol Great! Thanks a lot! As for that indentation, I just had a look at macOS system implementations and the indentation is used in very rare variety of cases, so I will probably remove it.

Here are few examples from system implementation:
![screen shot 2018-10-17 at 11 46 48 am](https://user-images.githubusercontent.com/1524024/47078163-f5c69580-d202-11e8-8f16-7cd8867ae5db.png)
![screen shot 2018-10-17 at 11 48 03 am](https://user-images.githubusercontent.com/1524024/47078165-f5c69580-d202-11e8-9165-945af3f6d134.png)

Also, do you think we need to add available languages title? I think it will be more consistent that way. What do you think?

jansol commented:
Yeah looks like not indenting is more in line with the rest of the system. (And we can always put the indentation back if it turns out problematic)

I don't think a subtitle for the available languages is necessary - the menu entry that opens this menu is already called `Language` so that seems quite obvious...

cmyr commented:
I agree that the indentation might not be necessary.

I've played around with this, and I do have a couple notes:

- Ideally the list would always be the five most recently used languages. However, if I select a language that's already recently used, its state isn't updated; so if I select the fifth most recently used language, and then select a new language, the language I was just using is removed from the list. My solution might be to always have the recents list be in descending order, most recently used to last?

- I was able at some point to get a language called 'Language' added to the recents list. This was on first launch. I'm not sure how I did this. 😕

Overall this looks good, and feels like a nice improvement; I look forward to getting it merged. 👍

demalex commented on XiEditor/Storage/KeyValueStorage.swift:
```diff
@@ -0,0 +1,23 @@
+// Copyright 2016 The xi-editor Authors.
```
@cmyr 😆 You are right. I have just copy pasted the header. We might need to have custom header and automatically add it to all newly created source files. What do think?

cmyr replied:
This is an Xcode feature? I'd forgotten about that, it could be helpful though!

jansol replied:
It is.

demalex commented on XiEditor/Storage/KeyValueStorage.swift:
```diff
@@ -0,0 +1,23 @@
+
+import Foundation
+
+protocol KeyValueStorage {
```
@cmyr I will add a comment 👍🏻

demalex commented:
@cmyr I just played with the application and feature for a while and did not have a change to reproduce the issue with initial `Language` value. Could you please provide more details?

As for descending ordering I think it would be better to have it. So, I will apply change and let you know when it is ready! Thanks.

demalex commented:
@cmyr I have updated implementation. Please, have a look at it! Now most recent language appears at the top of the list and pushes previously used languages to the bottom. If one of the recently used languages is selected it is moved to the top of the list again. Thus, recently used languages list represents ordered history of languages which were used.

demalex commented:
@cmyr Yes, I can see the issue. It appeared because I have changed Language menu building logic a bit. 
From what I see, when application started it receives `selected languages` and `available languages` notifications. Bu, when main window is closed and then opened again only `selected language` notification is received. Could you please tell, whether it is right behavior?

cmyr commented:
@demalex Ah, right, this makes sense.

What we should do is handle `available_languages` in the app delegate, and then make it accessible from there to all view controllers, instead of trying to handle it in the view controller itself.

basically: `available_languages` is really a global setting, but `selected_language` is scoped to the individual document.

demalex commented:
@cmyr Initially I had same thought, but then I decided to clarify this behaviour. Thanks for the explanation! 

Basically I renamed `SettingsStore` to `LanguagesStore` and now it can provide recently used languages as well as available languages. I think `Settings` it is another entity and it requires separate abstraction to be created (it could be very similar to `LanguagesStore`).

I do not think that it is a good idea to keep all shared objects in `AppDelegate` and get access them by calling to shared application and then make a cast to `AppDelegate`. Such approach violates design rules and makes code hard to maintain, change and test. I would rather pass dependencies explicitly, but it requires code to be refactored and such improvement out of this feature scope. 

Nevertheless, I have updated the implementation. Please have a look and let me know what you think.

Thanks!

mmatoszko commented on XiEditor/Storage/LanguagesStore.swift:
```diff
@@ -0,0 +1,63 @@
+        
+        let newLanguage = language.trimmingCharacters(in: .whitespacesAndNewlines)
+        
+        if newLanguage.count == 0 {
```
This check seems like a perfect job for a `guard` 😄 
`guard newLanguage.count > 0 else {
    return
}`

demalex replied:
@mmatoszko you think `if` can't handle it? 😎 It is not a big difference in this situation what to use. I think it easy to read simple `if` than `guard`. `guard` is good when you need to declare variables but do not want to increase cyclomatic complexity. Yes, compiler checks and if there is no `return` at some point it breaks compilation, but in such case (having tests) it is not so critical.

mmatoszko replied:
totally agree :+1:

mmatoszko commented on XiEditorTests/Mocks/DictionaryStorage.swift:
```diff
@@ -0,0 +1,36 @@
+
+class DictionaryStorage: KeyValueStorage {
+
+    var dictionary: [String : Any]
```
why not private?

demalex replied:
@mmatoszko `DictionaryStorage` should be used for tests when you need to mock object that conforms to `KeyValueStorage` protocol. It is added to Tests target only.

cmyr commented:
@demalex I definitely agree that our use of app delegate is pretty sketchy, and I'm open to refactoring. One related project is #275, which would move most of the RPC handling stuff out of the app delegate as well. 

Have been a bit distracted the past few days, but will try to give this another proper review either tonight or tomorrow. Thanks for your patience!

demalex commented:
@cmyr Thanks a lot and take your time! Let me know if you think that something needs to be improved.   I will have a look at another task for now.

demalex commented:
@cmyr Thanks a lot and take your time! Let me know if you think that something needs to be improved.   I will have a look at another task for now.

cmyr commented on XiEditor/EditViewController.swift:
```diff
@@ -133,11 +133,10 @@ class EditViewController: NSViewController, EditViewDataSource, FindDelegate, Sc
+    private var currentLanguage: String = ""
+
+    /// Object provides available and recently used languages
+    var languagesProvider: LanguagesProvider!
```
Okay, so I've been thinking about this a bit:

I think it's important that we have a clear sense of what the source of truth is for language information. I'm thinking of the situation where (theoretically) a new plugin is loaded in core that adds a new language: What happens in this case? As currently implemented, the `LanguagesProvider` in the `AppDelegate` would get updated, but the individual views wouldn't end up being notified?

There are a variety of approaches to this. One simple one would be to have all access to the `languagesProvider` be explicitly shared; the view controller would be returning a reference to the document, which would be returning the shared object in the AppDelegate. Then when the state changed, we would iterate through the view controllers and notify them of the state change, and they can reload menus etc as needed.

Alternatively, when the state gets set in core we could go and set the new `languageProvider` on each view controller, and then handle any updates in a `didSet` method.

Another confusion here is that because `LanguagesStore` is a class, everyone should currently be sharing a single instance, and so the state _should_ be in sync; but `LanguagesProvider` isn't declared as a class protocol, so if it were implemented as a struct this would no longer be true?

Another thing I might think about is whether this state belongs more in the document or in the view controller. I think it might be reasonable to have it only be in the document? I could go either way, though.

demalex replied:
@cmy Thank you for review! 

You are right, class protocol constraint definitely needs to be added to `LanguagesProvider` 👍🏻 

Ideally `ViewController` should not contain any state. ViewController is an entity of UI / Presentation layer and its main responsibility to be a glue between model (view-model) and view and handle view's lifecycle. I definitely would not keep state in view controller and most likely it is Document object's responsibility.

I have been thinking about it a bit and I came up with possible solution for the problem. The idea is to have `LanguagesStore` as single source of truth and listeners which can listen to notifications about the updates. The `LanguageStore` itself can be updated by receiving notification from the core. I have created high level UML diagram to make it a little bit clear. Please, have a look:

![languages store-2](https://user-images.githubusercontent.com/1524024/47388232-f636c300-d711-11e8-8895-d469f048a10d.png)

Diagram is high level in some cases (Document and XiDocumentController, Core), but still brings visibility how data flows from presentation layer to core and to `LanguagesStore`.

What do you think?

cmyr replied:
So I think we're broadly on the same page; what we want is a single source of truth and a notification mechanism.

We have options in terms of what our notification mechanism is. The simplest is that when the state changes, we iterate through our documents and call some `stateChanged()` function; then each document goes and checks the new state and performs some update logic. Another reasonable option is to lean on the Cocoa `NotificationCenter` infrastructure, which you're doing here. In _this_ case, though, I think it makes more sense to pass some reference type containing the state along as the notification's object; so in this case I would register for, say, the 'availableLanguagesChanged' notification, and the notification would contain a reference to some type, say, `CoreState`, (which can be a protocol) which has a member `var availableLanguages { get }`.

I think this is mostly what you're suggesting? The main thing I want to push is that we'd like to balance the benefits (clear separation of responsibilities, clear flow of data) against the costs of increasing architectural complexity; so whatever we do here should be something that we anticipate being the general model for communicating changes to shared state.

demalex replied:
@cmyr Yes, its true! This is what I am suggesting. Please, take a note that `LanguageStoringNotification` is actually an interface. Interface name is misleading and will be changed. Having interface makes a loose coupling and allows to do checking during compile time, while `NSNotification.Name` is actually a `String` and it is easy to break an implementation. So, I would not use `NotificationCenter`.

So, the basic idea is very simple: 

1. Subscriber implements protocol and be subscribed for receiving updates from `LanguagesStore`
2. `LanguagesStore` receives updates from the `Core`, updates it's state and notifies all subscribers by iterating over the collection and calling appropriate function
3. As you mentioned above, updated state of the store (as a type of lets say `LanguageStoringOutput`) can passed to the function as parameter

What do you think? Does it make sense?

Architecture is what allows you to manage complexity and makes it easy to adapt the code base for new requirements. So, there is always some level of architecture complexity. But, I agree, there should be a trade off.

Please, let know what do you think about my suggestion. Thank you!

cmyr replied:
Okay, thanks for the explanation. I think we're on the same page, and this approach sounds reasonable. 👍

cmyr replied:
My one thought is that we might want to eventually expand this so that we have a single mechanism for various types of state (themes are an obvious example), so I'd keep that in mind; for instance we might want multiple types of events that can be subscribed to, _or_ we might want them to be a package deal.

demalex replied:
@cmyr Thank you very much for the feedback! Yes, I am also thinking about generic solution that can be used in variety of cases. I was busy at work these days and going to have look at it on the weekend. Thanks!

cmyr commented on XiEditor/Storage/LanguagesStore.swift:
```diff
@@ -0,0 +1,63 @@
+protocol LanguagesProvider {
+    func fetchRecentlyUsedLanguages() -> [String]
+    func saveRecentlyUsed(language: String)
+    var availableLanguages: [String] { get set }
```
I'm not sure if we really want the `set` method exposed here? `availableLanguages` should only really be set in response to a core RPC, so exposing the setter to other users feels a bit iffy.

demalex replied:
@cmyr You are absolutely right. I think we need to change data flow a bit. I have some ideas and will share it soon.

cmyr commented:
@demalex just checking in, if this is ready for a review just :bell: me!

demalex commented:
@cmyr Sorry for the delay in answering. I am busy at work lately, but I am moving on with implementation. I hope, I can finish it soon. Let me know if it acceptable.

cmyr commented:
@demalex absolutely, work at whatever pace is good for you. I'm generally more worried that things are blocked on review or some question I've missed, so wanted to make sure that wasn't the case.

cmyr commented:
@demalex are you still making progress on this? It feels like it's really close, and it would be nice to get this feature done, but no worries if you're busy. 😃

demalex commented:
@cmyr Hi! Excuse me that it is taking so long to complete this. I a little bit overloaded at work at the end of the year, but I am making a progress and submit a new patch set soon. Hope it works for you.
Happy holidays! 🎄

cmyr commented:
@demalex amazing, glad to hear you're still hacking away. Happy holidays!

demalex commented:
Closed it due to big difference between master and feature branch. Will resubmit it soon

jansol commented:
@demalex ping?
